File: examples/01-intro-c/example8.c
// =================================================================
//
// File: example8.c
// Author(s): 
//					A01651517 Pedro Luis González Roa
//					A01703947 Juan Alejandro Alcántara Minaya
// Description: This file contains the code that implements the
//				enumeration sort algorithm. The time this implementation
//				takes ill be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"

#define SIZE 10000

// implement your code
void enum_sort(int* a, int* pos, int* result) {
	for (int i = 0; i < SIZE; i++) {
		pos[i] = 0;
		for (int j = 0; j < SIZE; j++) {
			if (a[i] > a[j] || a[i] == a[j] && i > j) {
				pos[i] += 1;
			}
		}
	}
	for (int i = 0; i < SIZE; i++) {
		result[pos[i]] = a[i];
	}
}

int main(int argc, char* argv[]) {
	int i, *a, *result, *pos;
	double ms;

	a = (int*) malloc(sizeof(int) * SIZE);
	pos = (int*) malloc(sizeof(int) * SIZE);
	result = (int*) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("before", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		// call the implemented function
		enum_sort(a, pos, result);

		ms += stop_timer();
	}
	display_array("after", result);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	free(pos);
	free(result);
	return 0;
}

File: examples/04-java-threads/Example8.java
// =================================================================
//
// File: Example8.java
// Author: Pedro González A01651517, Juan Alcantara A01703947
// Description: This file implements the enumeration sort algorithm.
// 				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;

public class Example8 extends Thread {
	private static final int SIZE = 100_000;
	private static final int MAXTHREADS = Utils.MAXTHREADS - 1;

	private int start;
	private int array[];
	private int res[];

	public Example8(int start, int array[], int res[]) {
		this.start = start;
		this.array = array;
		this.res = res;
	}

	public void run() {
		int n = 0;
		for (int i = start; i < SIZE; i += MAXTHREADS) {
			n = 0;
			for (int j = 0; j < SIZE; j++) {
				if (array[i] > array[j] || array[i] == array[j] && i > j) {
					n++;
				}
			}
			res[n] = array[i];
		}
	}
	
	// place your code here
	public static void main(String args[]) {
		long startTime, stopTime;
		double ms;


		// Create random array
		int array[] = new int[SIZE];
		// Create response array
		int res[] = new int[SIZE];

		// Create threads array
		Example8 threads[] = new Example8[MAXTHREADS];

		ms = 0;
		System.out.println("Starting testing...");
		for (int j = 1; j <= Utils.N; j++) {
			Utils.randomArray(array);
			if (j == Utils.N) {
				Utils.displayArray("before", array);
			}

			// Initialize threads
			for (int i = 0; i < MAXTHREADS; i++) {
				threads[i] = new Example8(i, array, res);
			}
			startTime = System.currentTimeMillis();
			// Start threads
			for (int i = 0; i < MAXTHREADS; i++) {
				threads[i].start();
			}
			// Stop threads
			for (int i = 0; i < MAXTHREADS; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
			
			// Get last result
			if (j == Utils.N) {
				Utils.displayArray("after", res);
			}
		}
		System.out.printf("avg time = %.5f ms\n", (ms / Utils.N));
	}
}
